pymc4/gp/cov.py

```diff
@@ -0,0 +1,162 @@
+        return self._kernel.matrix(X1, X2, **kwargs)
+
+    def evaluate_kernel(self, X1, X2, **kwargs):
+        """Evaluate kernel at certain points
+
+        Parameters
+        ----------
+        X1 : tensor, array-like
+            First point(s)
+        X2 : tensor, array-like
+            Second point(s)
+        """
+        return self._kernel.apply(X1, X2, **kwargs)
+
+    def __add__(self, cov2):
+        return CovarianceAdd(self, cov2)
+
+    def __mul__(self, cov2):
+        return CovarianceProd(self, cov2)
+
+    @property
+    def feature_ndims(self):
+        return self._feature_ndims
+
+
+class Combination(Covariance):
+    r"""Combination of two or more covariance functions
+
+    Parameters
+    ----------
+    cov1 : pm.gp.Covariance
+        First covariance function.
+    cov2 : pm.gp.Covariance
+        Second covariance function.
+    """
+
+    def __init__(self, cov1, cov2, **kwargs):
+        self.kernel1 = cov1._kernel
+        self.kernel2 = cov2._kernel
+        if cov1.feature_ndims != cov2.feature_ndims:
+            raise ValueError("Cannot combine kernels with different feature_ndims")
+        super().__init__(cov1.feature_ndims, **kwargs)
+
+
+class CovarianceAdd(Combination):
+    r"""Addition of two or more covariance functions.
+
+    Parameters
+    ----------
+    feature_ndims : int
+        number of rightmost dims to include in kernel computation
+    """
+
+    def _init_kernel(self, feature_ndims, **kwargs):
+        return self.kernel1 + self.kernel2
+
+
+class CovarianceProd(Combination):
+    r"""Product of two or more covariance functions.
+
+    Parameters
+    ----------
+    feature_ndims : int
+        number of rightmost dims to include in kernel computation
+    """
+
+    def _init_kernel(self, feature_ndims, **kwargs):
+        return self.kernel1 * self.kernel2
+
+
+class Stationary(Covariance):
+    r"""Base class for all Stationary Covariance functions"""
+
+    @property
+    def length_scale(self):
+        r"""Length scale of the covariance function"""
+        return self._length_scale
+
+
+class ExpQuad(Stationary):
+    r"""Exponentiated Quadratic Stationary Covariance Function.
+    Commonly known as the Radial Basis Kernel Function.
+
+    .. math::
+       k(x, x') = \sigma^2 \mathrm{exp}\left[ -\frac{(x - x')^2}{2 l^2} \right]
+
+    where :math:`\sigma` = ``amplitude``
+          :math:`l` = ``length_scale``
+
+    Parameters
+    ----------
+    amplitude : tensor, array-like
+        The :math:`\sigma` parameter of RBF kernel, amplitude > 0
+    length_scale : tensor, array-like
+        The :math:`l` parameter of the RBF kernel
+    feature_ndims : int, optional
+        number of rightmost dims to include in kernel computation
+    kwargs : optional
+        Other keyword arguments that tfp's ``ExponentiatedQuadratic`` kernel takes
+    """
+
+    def __init__(self, amplitude, length_scale, feature_ndims=1, **kwargs):
+        self._amplitude = amplitude
```

Comment:
The reason I didn't include an amplitude or scale parameter in the covariance functions in PyMC3 is because it clutters the notation for product kernels.  The user has to fix all the amplitudes at 1.0 except one of them or else it's overparameterized.  

    cov = cov1(amplitude=eta**2, ...) * cov2(amplitude=1.0, ...) * cov3(amplitude=1.0)

vs.

    cov = eta**2 * cov1(...) * cov2(...) * cov3(...)

I can see how it's cleaner than having to multiply a single cov by some scalar parameter though, so I think it would be best to either not have the amplitude parameter or use a default amplitude of 1.0.

Reply:
True! Will have to add a ``__rmul__`` and ``__radd__`` methods to do that + some tests. I will do this by today!

Reply:
This is a little tricky unless tensorflow_probability supports white noise kernel. I have made a request to implement the white noise kernel in tensorflow_probability [tensorflow/probability#852](https://github.com/tensorflow/probability/issues/852). Without this, I will have to drop the ``_kernel`` object and implement the call method to call the individual covariance functions individually. Doesn't seem that bad but it would be nice if tfp implemented a white noise kernel.

Reply:
@bwengals I have added the support for ``__radd__`` and ``__rmul__``. I will add tests once everyone agrees with the current design, If tensorflow_probability ever adds the white noise kernel then this design can be easily refactored to take that into account. But temporarily, this is what I have ended up with...

Reply:
Take your time!  

Right I see, you do need an amplitude parameter without `__radd__` and `__rmul__`.  How come a white noise kernel is needed?

Reply:
Oh, sorry! I wanted to say the constant kernel. Actually, I have designed the covariance class such that the ``_init_kernel`` method initializes a kernel which is an instance of tfp's ``PositiveSemidefiniteKernel`` class. But currently, tfp doesn't support combination of covariance with non-covariance functions. So, we would need something like a constant kernel in which we can wrap the non-covariance kernel and use the ``_init_kernel`` method (and ``__call__`` method) to take care of everything else.

Reply:
Oh okay I see what you're saying now, makes sense.

Reply:
It may make sense to implement the ConstantKernel and use that for amplitude. It may make it clearer to users how the amplitude works, and would avoid trying to specify multiple amplitudes for multiplicative kernels. What do you think?